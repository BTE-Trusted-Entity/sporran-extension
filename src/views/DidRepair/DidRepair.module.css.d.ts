export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const fee: string;
export const info: string;
export const buttonsLine: string;
export const cancel: string;
export const submit: string;
export const errorTooltip: string;
