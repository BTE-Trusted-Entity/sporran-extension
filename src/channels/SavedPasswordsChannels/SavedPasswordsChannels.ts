import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
