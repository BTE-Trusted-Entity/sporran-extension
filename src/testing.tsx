export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
