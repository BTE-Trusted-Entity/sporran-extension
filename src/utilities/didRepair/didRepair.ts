import BN from 'bn.js';
import {
  Blockchain,
  ConfigService,
  Did,
  DidUri,
  KiltEncryptionKeypair,
  KiltKeyringPair,
  SignExtrinsicCallback,
  SubmittableExtrinsic,
  Utils,
} from '@kiltprotocol/sdk-js';

import {
  deriveEncryptionKeyFromSeed,
  getIdentityCryptoFromSeed,
} from '../identities/identities';
import { getDidEncryptionKey, getFullDidDocument } from '../did/did';
import { makeFakeIdentityCrypto } from '../makeFakeIdentityCrypto/makeFakeIdentityCrypto';

async function getSignedTransaction(
  keypair: KiltKeyringPair,
  sign: SignExtrinsicCallback,
  fullDid: DidUri,
  newKey: KiltEncryptionKeypair,
): Promise<SubmittableExtrinsic> {
  const fullDidDocument = await getFullDidDocument(fullDid);

  const existingKey = getDidEncryptionKey(fullDidDocument);

  const keyToUpdate =
    '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e';
  const noUpdateNeeded =
    Utils.Crypto.u8aToHex(existingKey.publicKey) !== keyToUpdate;
  if (noUpdateNeeded) {
    throw new Error('DID repair not needed');
  }

  const api = ConfigService.get('api');

  const authorized = await Did.authorizeBatch({
    batchFunction: api.tx.utility.batchAll,
    did: fullDid,
    extrinsics: [
      api.tx.did.removeKeyAgreementKey(Did.resourceIdToChain(existingKey.id)),
      api.tx.did.addKeyAgreementKey(Did.publicKeyToChain(newKey)),
    ],
    submitter: keypair.address,
    sign,
  });

  return authorized.signAsync(keypair);
}

export async function getFee(did: DidUri): Promise<BN> {
  const { keypair, sign, fakeSeed } = makeFakeIdentityCrypto();
  const newKey = deriveEncryptionKeyFromSeed(fakeSeed);
  const extrinsic = await getSignedTransaction(keypair, sign, did, newKey);

  return (await extrinsic.paymentInfo(keypair)).partialFee;
}

const currentTx: Record<string, SubmittableExtrinsic> = {};

export async function sign(did: DidUri, seed: Uint8Array): Promise<string> {
  const { sign, keypair } = await getIdentityCryptoFromSeed(seed);
  const newKey = deriveEncryptionKeyFromSeed(seed);
  const extrinsic = await getSignedTransaction(keypair, sign, did, newKey);

  const hash = extrinsic.hash.toHex();
  currentTx[hash] = extrinsic;
  return hash;
}

export async function submit(hash: string): Promise<void> {
  const extrinsic = currentTx[hash];
  await Blockchain.submitSignedTx(extrinsic);
  delete currentTx[hash];
}
