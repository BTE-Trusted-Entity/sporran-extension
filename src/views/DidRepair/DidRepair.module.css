.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.fee {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 1rem 0 0;
}

.info {
  composes: subline from '../../components/Typography/Typography.module.css';
  white-space: pre-line;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin: 1.5rem 0 0;
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
