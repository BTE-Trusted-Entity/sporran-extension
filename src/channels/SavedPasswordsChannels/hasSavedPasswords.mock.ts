import { act } from '@testing-library/react';

import { hasSavedPasswordsChannel } from './SavedPasswordsChannels';

jest.mock('./SavedPasswordsChannels');
const hasSavedPasswordsPromise = Promise.resolve(false);
(hasSavedPasswordsChannel.get as jest.Mock).mockReturnValue(
  hasSavedPasswordsPromise,
);

export async function waitForHasSavedPasswords(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}
