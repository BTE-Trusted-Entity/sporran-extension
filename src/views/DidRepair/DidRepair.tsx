import { FormEvent, useCallback, useState } from 'react';
import { Link } from 'react-router-dom';
import BN from 'bn.js';
import { browser } from 'webextension-polyfill-ts';

import { DidUri } from '@kiltprotocol/sdk-js';

import * as styles from './DidRepair.module.css';

import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { Identity } from '../../utilities/identities/types';
import { getIdentityDid } from '../../utilities/identities/identities';
import {
  PasswordError,
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { getFee, sign, submit } from '../../utilities/didRepair/didRepair';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { useAddressBalance } from '../../components/Balance/Balance';
import { KiltCurrency } from '../../components/KiltCurrency/KiltCurrency';
import { asKiltCoins } from '../../components/KiltAmount/KiltAmount';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import { paths } from '../paths';

interface Props {
  identity: Identity;
}

function useCosts(
  address: string,
  did: DidUri,
): {
  fee?: BN;
  error: boolean;
} {
  const fee = useAsyncValue(getFee, [did]);

  const balance = useAddressBalance(address);
  const error = Boolean(balance && fee && balance.transferable.lt(fee));

  return { fee, error };
}

export function DidRepair({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { address } = identity;
  const did = getIdentityDid(identity);

  const { fee, error } = useCosts(address, did);
  const [txHash, setTxHash] = useState<string>();

  const [submitting, setSubmitting] = useState(false);
  const [status, setStatus] = useState<'pending' | 'success' | 'error' | null>(
    null,
  );

  const passwordField = usePasswordField();

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      try {
        const { seed } = await passwordField.get(event);

        setSubmitting(true);
        setStatus('pending');

        const hash = await sign(did, seed);
        setTxHash(hash);

        await submit(hash);

        setStatus('success');
      } catch (error) {
        if (error instanceof PasswordError) {
          return;
        }
        setSubmitting(false);
        setStatus('error');
      }
    },
    [did, passwordField],
  );

  const closeModal = useCallback(() => {
    setStatus(null);
  }, []);

  if (!fee) {
    return null; // blockchain data pending
  }

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>{t('view_DidRepair_heading')}</h1>
      <p className={styles.subline}>{t('view_DidRepair_subline')}</p>

      <IdentitySlide identity={identity} />

      <p className={styles.fee}>
        {t('view_DidRepair_fee')}
        {asKiltCoins(fee, 'costs')} <KiltCurrency />
      </p>

      <p className={styles.info}>{t('view_DidRepair_info')}</p>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button
          type="submit"
          className={styles.submit}
          disabled={submitting || error}
        >
          {t('common_action_sign')}
        </button>
        <output
          className={styles.errorTooltip}
          hidden={!error || Boolean(status)}
        >
          {t('view_DidRepair_insufficientFunds', [asKiltCoins(fee, 'costs')])}
        </output>
      </p>

      {status && (
        <TxStatusModal
          identity={identity}
          status={status}
          txHash={txHash}
          onDismissError={closeModal}
          messages={{
            pending: t('view_DidRepair_pending'),
            success: t('view_DidRepair_success'),
            error: t('view_DidRepair_error'),
          }}
        />
      )}

      <LinkBack />
      <Stats />
    </form>
  );
}
