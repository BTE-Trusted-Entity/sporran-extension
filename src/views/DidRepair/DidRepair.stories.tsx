import { Meta } from '@storybook/react';

import { legacyIdentity } from '../../utilities/identities/IdentitiesProvider.mock';

import { DidRepair } from './DidRepair';

export default {
  title: 'Views/DidRepair',
  component: DidRepair,
} as Meta;

export function Template(): JSX.Element {
  return <DidRepair identity={legacyIdentity} />;
}
