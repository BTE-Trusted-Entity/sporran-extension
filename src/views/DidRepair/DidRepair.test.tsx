import { BalanceUtils } from '@kiltprotocol/sdk-js';

import { identitiesMock as identities, render } from '../../testing/testing';

import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import { getFee } from '../../utilities/didRepair/didRepair';

import { DidRepair } from './DidRepair';

jest.mock('../../utilities/didRepair/didRepair', () => ({
  getFee: jest.fn(),
}));
jest.mocked(getFee).mockResolvedValue(BalanceUtils.toFemtoKilt(0.01));

describe('DidRepair', () => {
  it('should render', async () => {
    const { container } = render(
      <DidRepair
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
