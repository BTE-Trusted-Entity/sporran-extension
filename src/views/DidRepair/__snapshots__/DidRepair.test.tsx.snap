// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidRepair should render 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      DID Update
    </h1>
    <p
      class="subline"
    >
      Please update your DID
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="fee"
    >
      Fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="info"
    >
      This transaction updates your DID.
The fee has already been refunded.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to update your DID
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
